src/main/java/br/com/swconsultoria/certificado/CertificadoService.java

```diff
@@ -278,4 +270,53 @@ public static Certificado getCertificadoByCnpjCpf(String cnpjCpf) throws Certifi
                         cnpjCpf));
     }
 
+    private static Protocol getProtocoloCertificado(final Certificado certificado, InputStream cacert) throws CertificadoException {
```

Comment:
Mantive o "padrão" de gerar o protocolo/SocketFactoryDinamico a cada chamada, mas é possível mudar a abordagem e usar o register, mantendo no próprio cache de `Protocol`. O ponto "negativo" é que se for uma aplicação web ficaria em "cache pra sempre", para resolver isso poderia implementar uma opção para remover do cache que poderia ser utilizada, por exemplo, ao editar o cadastro do certificado do cliente.

Reply:
É, aqui mora o perigo. Temos que garantia que cada chamada seja unica e pegue o certificado daquela Thread/execuçao, um cache iria por isso em risco.
Talvez se conseguirmos identicar o protocol com um hash do certificado (Algo que identifique a troca de certificado) por exemplo, poderiamos saber que ele pertence ao mesmo do atual e retornar o já ativo. Talvez um Objeto statico que renovariamos diariamente... Só precisamos ter cuidado de não estar querendo alterar algo simples por nada. Essa modificaçao traria mais velocidade, porém nunca tive reclamacao do mesmo.

Reply:
Na verdade quando us o modo multithreading o consumidor deve usar o método `getHttpsClient` e com isso já tera um certificado/protocolo "único/isolado" na conexão (como foi feito na `DistribuicaoDFe` usando o `StubUtil`) 

Como a performance não é um problema, vou manter como está sem implementar um cache ou algo assim.
Da forma como está é como testei e como o Bruno S. também já testou/homologou.